Refactor liveness probing

timstclair opened the pull request:
This commit builds on the work from https://github.com/kubernetes/kubernetes/pull/14221 and creates an independent worker for every liveness probe. Liveness probes behave largely the same as readiness probes, so much of the code is shared by introducing a `probeType` paramater to distinguish  the type when it matters. The circular dependency between the runtime and the prober is broken by exposing a shared liveness [`ResultsManager`](https://github.com/kubernetes/kubernetes/pull/15274), owned by the kubelet. Finally, an `Updates` channel is introduced to the ResultsManager so the kubelet can react to unhealthy containers immediately (this will be reused for setting readiness state in a following PR).

~~Depends on https://github.com/kubernetes/kubernetes/pull/15274 (note that I've included this)~~

Prerequisite for https://github.com/kubernetes/kubernetes/issues/12866

cc/ @kubernetes/goog-node

timstclair commented:
Rebased.

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -1971,6 +1978,12 @@ func (kl *Kubelet) syncLoopIteration(updates <-chan kubetypes.PodUpdate, handler
+		// We only care about failures (signalling container death) here.
+		if update.Result == proberesults.Failure {
+			glog.V(1).Infof("SyncLoop (container unhealthy).")
+			handler.HandlePodSyncs(kl.podManager.GetPods())
```
I don't think we should sync all the pods here. Can we map the container to the corresponding pod and only sync that pod?

timstclair replied:
Done. I added the pod object to the update. It's a less elegant API, but I think it's simpler than trying to map the container ID back to the pod.

yujuhong commented on pkg/kubelet/prober/manager.go:
```diff
@@ -74,36 +80,63 @@ type manager struct {
+// Type of probe (readiness or liveness)
+type probeType bool
+
+const (
```
nit: I understand bool is more economical here, but using int seems more extensible and safer (e.g.,  accidentally used in an if-else condition)
```
const (
    liveness probeType = iota // 0
    readiness                 // 1
)
```

timstclair replied:
Done. Good point about the if-else.

yujuhong commented on pkg/kubelet/prober/results/results_manager_test.go:
```diff
@@ -18,11 +18,14 @@ package results
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 )
 
+const timeout = 30 * time.Second
```
Brendon advocated to use `util.ForeverTestTimeout` for all test timeout like this.

timstclair replied:
Done.

yujuhong commented on pkg/kubelet/prober/results/results_manager_test.go:
```diff
@@ -41,3 +44,57 @@ func TestCacheOperations(t *testing.T) {
+		// still fail.
+		select {
+		case u := <-m.Updates():
+			t.Errorf("Unepected update %v: %s %s", u, msg)
```
typo: Unexpected.
also it's missing one variable.

timstclair replied:
Done.

yujuhong commented on pkg/kubelet/prober/results/results_manager_test.go:
```diff
@@ -41,3 +44,57 @@ func TestCacheOperations(t *testing.T) {
+	}
+
+	expectNoUpdate := func(msg string) {
+		// NOTE: Since updates are accumulated assynchronously, this method is not guaranteed to fail
```
s/assynchronously/asynchronously/

timstclair replied:
Done.

yujuhong commented on pkg/kubelet/prober/worker.go:
```diff
@@ -53,13 +55,22 @@ type worker struct {
+	probeType probeType) *worker {
+
+	var spec *api.Probe
+	if probeType == readiness {
```
nit: use switch instead?

timstclair replied:
Done.

yujuhong commented on pkg/kubelet/prober/worker_test.go:
```diff
@@ -72,135 +73,160 @@ func TestDoProbe(t *testing.T) {
-	if _, ok := m.readinessProbes[containerPath{podUID, containerName}]; ok {
-		t.Error("Expected worker to be cleared.")
+		if ready, _ := m.resultsManager(probeType).Get(containerID); !ready {
+			t.Fatalf("[%s] Expected result to be true.", probeType)
```
I think you meant to print `probeType.String()`? The same applies to the `t.Errorf` messages below.

timstclair replied:
The `.String()` method is automatically called in these places. I think it behaves kind of like java's `toString()`

yujuhong replied:
I see. I remembered and then I forgot.

yujuhong commented on pkg/kubelet/prober/worker_test.go:
```diff
@@ -232,12 +258,20 @@ func getTestPod(probeSpec api.Probe) api.Pod {
-func (f CrashingProber) ProbeReadiness(_ *api.Pod, _ api.PodStatus, c api.Container, _ kubecontainer.ContainerID) (probe.Result, error) {
-	panic("Intentional ProbeReadiness crash.")
+// Format probe type.
+func (t probeType) String() string {
```
Would this be useful in the non-test code? Maybe we should move this to probe/manager.go.

timstclair replied:
I don't think it's currently being used outside of test code, but it could be handy for debugging in the future.

yujuhong commented on pkg/kubelet/prober/results/results_manager.go:
```diff
@@ -87,3 +100,21 @@ func (m *manager) Remove(id kubecontainer.ContainerID) {
 }
+
+func (m *manager) Updates() <-chan Update {
+	if m.updates == nil {
```
It's a bit concerning that `m.updates` can be accessed/modified by a different goroutine without lock acquiring.

timstclair replied:
Could point. I'll push this to creation time.

timstclair replied:
I decided I preferred the current API to trying to push this to the constructor, so I made updates thread-safe. It's a bit messy though, so let me know if you have a better idea.

yujuhong commented on pkg/kubelet/prober/results/results_manager.go:
```diff
@@ -87,3 +100,21 @@ func (m *manager) Remove(id kubecontainer.ContainerID) {
+		select {
+		case m.updates <- update:
+		default:
+			// Don't block
```
What happens to dropped updates?

timstclair replied:
The idea was that the updates were just triggering a sync, so they would only be dropped if a sync was already happening (in which case it was unnecessary). Now that I'll only be syncing individual pods, I think it makes sense for the worker to just block at this point.

yujuhong replied:
Ack. I think the performance optimization (avoid syncing all pods) is necessary as we scale.

yujuhong commented on pkg/kubelet/prober/worker.go:
```diff
@@ -46,6 +45,9 @@ type worker struct {
 
+	// The type of the worker.
+	probeType probeType
+
```
nit: There are a several places in this file where you check the probeType in the if-else condition, but the rest of the code is agnostic to the probe type. Maybe you can remove all those conditions by setting them in the worker constructor? E.g., set the probe function to `m.prober.ProbeReadiness` if `probeType==readiness`, and vice versa. This makes the difference between two types of probes clear and it's easier to reuse the code. I am not sure if this is worth the efforts, so I'll leave it up to you to decide :)

timstclair replied:
I like having probeType stored on the worker, since I think it makes debugging a little easier. However, I could abstract away the 2 places I check the probeType:

- Create a prober.Probe method which takes the probeType and calls the correct version
- Create an initialValue function which returns the value to set during the initial delay

Alternatively, eliminating the probeType field from the worker would require adding fields for:

- resultsManager
- probeFunction
- initialValue

What are your thoughts on these approaches vs. what I have now?

yujuhong replied:
I am not against keeping probeType here, but would like to limit its use to debugging if possible. Adding fields seems cleaner to me because a worker is not going to change its probeType half way. I am ok with the current approach but would prefer adding fields. I wouldn't block the PR on this though.

timstclair replied:
Done (mostly - probeType is still passed to prober, I'll try to fix this in a later PR). Also made a few cleanup changes:

- prober.Manager is a worker field rather than method parameter
- removed side effect (`go w.run()`) from worker constructor for easier testing. This also exposed a bug in TestCleanUp which I fixed.

yujuhong commented on pkg/kubelet/prober/results/results_manager.go:
```diff
@@ -87,3 +115,26 @@ func (m *manager) Remove(id kubecontainer.ContainerID) {
+		// Verify updatse did not change while we were aquiring the lock
+		updates = m.updates.Load().(chan Update)
+		if updates == nil {
+			updates = make(chan Update)
```
Is it okay to not have any extra buffering here?

timstclair replied:
It's probably unnecessary, but I think a small buffer could probably smooth things out. I arbitrarily picked 20 - does that seem reasonable?

yujuhong replied:
sgtm. I think it'd be more than enough under regular circumstances. If it does become a bottleneck, that means something wrong in the sync loop itself and we probably need to fix that.

yujuhong commented on pkg/kubelet/prober/results/results_manager.go:
```diff
@@ -87,3 +115,26 @@ func (m *manager) Remove(id kubecontainer.ContainerID) {
 }
+
+func (m *manager) Updates() <-chan Update {
+	updates := m.updates.Load().(chan Update)
```
I understand the reason behind using the atomic value here, but was wondering whether we are over-optimizing...Anyway, lgtm.

timstclair replied:
Possibly, it does add some unfortunate complexity. Once I start using the Updates channel in the ReadinessManager I could just push its creation to the constructor and get rid of all this logic.

timstclair replied:
Removed atomic usage.

yujuhong commented:
lgtm but need to rebase.

timstclair commented:
Rebased & squashed. Reapplying LGTM.

timstclair commented:
Turns out go 1.3 doesn't have `atomic.Value`. Rather than messing with atomic pointers or mutexes, I decided to just simplify the whole thing and move the initialization to the constructor. It clutters the API a little, but greatly simplifies the code.

yujuhong commented:
> Turns out go 1.3 doesn't have atomic.Value

There is a custom type AtomicValue in pkg/util/atomic_value.go for this purpose, but I am okay either way :)

timstclair commented:
@k8s-bot test this

timstclair commented:
Reapplying LGTM after rebase & trivial test fix.

tallclair committed:
Refactor liveness probing

This commit builds on previous work and creates an independent
worker for every liveness probe. Liveness probes behave largely the same
as readiness probes, so much of the code is shared by introducing a
probeType paramater to distinguish the type when it matters. The
circular dependency between the runtime and the prober is broken by
exposing a shared liveness ResultsManager, owned by the
kubelet. Finally, an Updates channel is introduced to the ResultsManager
so the kubelet can react to unhealthy containers immediately.

timstclair commented:
Reapplying LGTM after fix to unit test.
